### DISCUSSION TITLE
help with multi string searches

### DISCUSSION THREAD

**Discussion by mooseyboots** [minad/corfu#512] [Q&A]
hi, thanks for this package.

i'm new to corfu (& orderless, vertico, embark, the whole family), so bear with me if this is old news. I've searched the repo for this issue, but not very well because it's unclear to me what to search for.

i'm struggling with multistring matching.

i know about `corfu-separator`, but i can't get it to work as i would like it to in elisp code.

my corfu config:

```lisp
(use-package corfu
  :config
  (global-corfu-mode)
  (setq corfu-auto t))
```

my orderless:

```lisp
(use-package orderless
  :ensure t
  :config
  ;; make spaces work:
  (setq completion-styles '(substring orderless basic)))
```

say i'm typing a call to a function in package `mastodon`:

```lisp
(masto-
```

and it has five words in its name all separated by hyphens. i know the last or second last one, but not the intermediate terms (the first of which names the subpackage the function is in, and maybe i've forgotten that, or i just forget the word order of the name).

if i change `corfu-separator` to a hyphen (char 45) rather than a space, i'd hope to be able to type say

```lisp
(masto-lastword
```

skipping the intermediate words. but it doesn't work, i get no candidates, whether i type `-` or insert it with `M-SPC`.

a concrete example:

```lisp
(masto-get
```

returns "no match". but mastodon.el is huge and has many `get` functions, it's just that none of them start with `masto-get`.

```lisp
(mastodon-get
```

ie. using the full package prefix, also doesn't work.

if i return the separator to the default space, and insert it using `M-SPC`, then it works:

```lisp
(masto get
```

is there a clash between the separator and the function characters? i can't grok why there would be. i would even have thought that splitting terms by both hyphen and space should work.

***

another example that i understand less well:

I type `C-x 8 RET`, then type `rh` only. The sole candidate is `RHINOCEROS`, but i'm looking for Greek rho. I type `o` to make `rho`, and only then do all the rho candidates appear. Why does `rh` not match them? It seems to be because no candidate names begin with the letters `rh`, but why is that how the query works? If type `C-x 8 RET M-SPC rh`, then it works. This is really tricky, because it makes you think there are no results, not just that they aren't displaying but that they don't exist... A more generous, less parsimonious, matching style would be in order here. Is it possible to configure one simply?

Thanks for any pointers.

**Comment by jdtsmith** [comment]
You have `substring` before orderless in `completion-styles`.  If the `substring` style produces matches, completion will _not_ continue on to `orderless`.  The first style to produce results wins; results are not combined.  

Give `(orderless basic)` a try.  You should also have a look at configuring orderless, as it can "act like" most other completion styles in situations you determine.

**Comment by mooseyboots** [comment]
thanks for the kind response. config of orderless it is!

removing substring from `completion-styles` and configuring `orderless-matching-styles` was what i needed to do.

apols for not knowing where to ask.
